File: clients/rust/src/mux.rs
// Copyright 2025 HEM Sp. z o.o.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// This file is part of an implementation of Ferrum Streaming Control Technology™,
// which is subject to additional terms found in the LICENSE-FSCT.md file.

use std::collections::HashMap;

use fsct::DeviceChangeEvent;
use futures::{SinkExt, StreamExt};
use log::warn;
use serde_json::{Value, Value as JsonValue, json};
use tokio::io::{AsyncRead, AsyncWrite};
use tokio::sync::{broadcast, mpsc, oneshot};
use tokio::sync::oneshot::Sender;
use tokio_util::codec::{FramedRead, FramedWrite, LinesCodec};
use uuid::Uuid;

use crate::rpc::{RpcNotification, RpcRequest, RpcResponse};

pub(crate) struct OutboundCall {
    pub(crate) method: String,
    pub(crate) params: JsonValue,
    pub(crate) reply: oneshot::Sender<Result<JsonValue, String>>,
}

type PendingMap = HashMap<u64, oneshot::Sender<Result<JsonValue, String>>>;

pub(crate) async fn run_mux<R, W>(
    mut reader: FramedRead<R, LinesCodec>,
    mut writer: FramedWrite<W, LinesCodec>,
    mut rx: mpsc::Receiver<OutboundCall>,
    device_tx: broadcast::Sender<DeviceChangeEvent>,
) where
    R: AsyncRead + Unpin,
    W: AsyncWrite + Unpin,
{
    let mut pending: PendingMap = HashMap::new();
    let mut next_id: u64 = 1;

    loop {
        tokio::select! {
            call = rx.recv() => {
                let Some(call) = call else { break };
                if handle_outbound(call, &mut writer, &mut pending, &mut next_id).await { break; }
            }
            line = reader.next() => {
                if handle_inbound(line, &mut pending, &device_tx) { break; }
            }
        }
    }
}

async fn handle_outbound<W: AsyncWrite + Unpin>(
    call: OutboundCall,
    writer: &mut FramedWrite<W, LinesCodec>,
    pending: &mut PendingMap,
    next_id: &mut u64,
) -> bool {
    let id = *next_id;
    *next_id += 1;
    let req = RpcRequest { jsonrpc: "2.0".into(), id: json!(id), method: call.method, params: call.params };
    match serde_json::to_string(&req) {
        Ok(line) => {
            if writer.send(line).await.is_err() {
                let _ = call.reply.send(Err("connection closed".into()));
                return true;
            }
            pending.insert(id, call.reply);
        }
        Err(e) => {
            let _ = call.reply.send(Err(format!("serialize error: {}", e)));
        }
    }
    false
}

fn handle_inbound(
    line_result: Option<Result<String, tokio_util::codec::LinesCodecError>>,
    pending: &mut PendingMap,
    device_tx: &broadcast::Sender<DeviceChangeEvent>,
) -> bool {
    let Some(line_result) = line_result else {
        fail_all_pending(pending, "connection closed");
        return true;
    };
    let line = match line_result {
        Ok(l) => l,
        Err(e) => {
            warn!("IPC read error: {}", e);
            fail_all_pending(pending, &format!("read error: {}", e));
            return true;
        }
    };
    dispatch_inbound(&line, pending, device_tx);
    false
}

fn fail_all_pending(pending: &mut PendingMap, reason: &str) {
    for (_, tx) in pending.drain() {
        let _ = tx.send(Err(reason.into()));
    }
}

fn dispatch_inbound(
    line: &str,
    pending: &mut HashMap<u64, oneshot::Sender<Result<JsonValue, String>>>,
    device_tx: &broadcast::Sender<DeviceChangeEvent>,
) {
    if let Ok(resp) = serde_json::from_str::<RpcResponse>(line) {
        handle_response(pending, resp);
    } else if let Ok(notif) = serde_json::from_str::<RpcNotification>(line) {
        handle_notification(&notif, device_tx);
    } else {
        warn!("IPC: unrecognized message: {}", &line[..line.len().min(200)]);
    }
}

fn handle_response(pending: &mut HashMap<u64, Sender<Result<Value, String>>>, resp: RpcResponse) {
    if let Some(id_u64) = resp.id.as_u64() {
        if let Some(tx) = pending.remove(&id_u64) {
            let result = resp.error.map(|e| Err(e.message)).unwrap_or_else(|| Ok(resp.result.unwrap_or(JsonValue::Null)));
            let _ = tx.send(result);
        }
    }
}

fn handle_notification(notif: &RpcNotification, tx: &broadcast::Sender<DeviceChangeEvent>) {
    if notif.method != "device_changed" {
        warn!("unknown notification: {}", notif.method);
        return;
    }
    let event_str = notif.params["event"].as_str().unwrap_or("");
    let device_id_str = notif.params["device_id"].as_str().unwrap_or("");
    let uuid = match Uuid::parse_str(device_id_str) {
        Ok(u) => u,
        Err(_) => {
            warn!("invalid device_id in notification: {}", device_id_str);
            return;
        }
    };
    let event = match event_str {
        "added" => DeviceChangeEvent::Added(uuid),
        "removed" => DeviceChangeEvent::Removed(uuid),
        _ => {
            warn!("unknown device_changed event: {}", event_str);
            return;
        }
    };
    let _ = tx.send(event);
}

File: clients/node/src/mux.ts
// Copyright 2025 HEM Sp. z o.o.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// This file is part of an implementation of Ferrum Streaming Control Technology™,
// which is subject to additional terms found in the LICENSE-FSCT.md file.

// Internal module: NDJSON read loop, request/response multiplexer, notification dispatch.

import { EventEmitter } from 'node:events';
import { createInterface } from 'node:readline';
import type { Socket } from 'node:net';
import {
  FsctError,
  isRpcNotification,
  isRpcResponse,
  MAX_LINE_BYTES,
  type RpcRequest,
} from './protocol.js';
import type { DeviceChangeEvent } from './types.js';

/** Default per-call timeout for Mux.call (ms). IPC calls hit in-memory state on the driver
 * and should return in milliseconds; anything past this points to a stuck driver. Pass 0
 * explicitly to disable. */
export const DEFAULT_CALL_TIMEOUT_MS = 5000;

interface PendingCall {
  resolve: (value: unknown) => void;
  reject: (reason: Error) => void;
  timer?: NodeJS.Timeout;
}

/**
 * Multiplexes JSON-RPC 2.0 requests/responses over a single socket and dispatches
 * server-sent notifications as EventEmitter events.
 *
 * Emits:
 * - 'deviceChanged' (DeviceChangeEvent) — device_changed notifications from the server
 * - 'error' (Error) — forwarded socket errors
 * - 'close' () — socket closed
 */
export class Mux extends EventEmitter {
  private readonly pending = new Map<number, PendingCall>();
  private readonly rl: ReturnType<typeof createInterface>;
  private nextId = 1;
  #destroyed = false;

  constructor(private readonly socket: Socket) {
    super();
    socket.on('error', (err: Error) => this.onError(err));
    socket.on('close', () => this.onClose());
    this.rl = createInterface({ input: socket, crlfDelay: Infinity });
    // Node.js readline re-emits socket errors on the Interface; suppress to avoid
    // double-handling since the socket 'error' listener below already handles them.
    this.rl.on('error', () => {});
    this.rl.on('line', (line: string) => this.dispatchLine(line));
  }

  get destroyed(): boolean {
    return this.#destroyed;
  }

  /** Send an RPC call and return a Promise that resolves with the result value.
   *  When `timeoutMs` is omitted, `DEFAULT_CALL_TIMEOUT_MS` is used. Pass `0` to disable. */
  call<T>(method: string, params: Record<string, unknown>, timeoutMs: number = DEFAULT_CALL_TIMEOUT_MS): Promise<T> {
    if (this.#destroyed) {
      return Promise.reject(new Error('IPC connection closed'));
    }

    return new Promise<T>((resolve, reject) => {
      const id = this.nextId++;
      const entry: PendingCall = {
        resolve: resolve as (value: unknown) => void,
        reject,
      };
      this.pending.set(id, entry);

      const request: RpcRequest = { jsonrpc: '2.0', id, method, params };
      const serialized = JSON.stringify(request);
      if (Buffer.byteLength(serialized, 'utf8') > MAX_LINE_BYTES) {
        this.pending.delete(id);
        reject(new Error('IPC request exceeds maximum line size'));
        return;
      }

      const line = serialized + '\n';

      this.socket.write(line, (err) => {
        if (err) {
          const cb = this.pending.get(id);
          if (cb) {
            this.pending.delete(id);
            if (cb.timer) clearTimeout(cb.timer);
            cb.reject(err);
          }
        }
      });

      if (timeoutMs > 0) {
        entry.timer = setTimeout(() => {
          const cb = this.pending.get(id);
          if (!cb) return;
          this.pending.delete(id);
          cb.reject(new Error(`IPC request timed out after ${timeoutMs}ms: method=${method}`));
        }, timeoutMs);
      }
    });
  }

  private dispatchLine(line: string): void {
    if (Buffer.byteLength(line, 'utf8') > MAX_LINE_BYTES) {
      this.failAll(new Error('IPC message exceeds maximum line size'));
      this.socket.destroy();
      return;
    }

    const trimmed = line.trim();
    if (!trimmed) return;

    let parsed: unknown;
    try {
      parsed = JSON.parse(trimmed);
    } catch {
      console.warn('[fsct-client] Received malformed JSON line, ignoring');
      return;
    }

    if (isRpcResponse(parsed)) {
      const id = typeof parsed.id === 'number' ? parsed.id : null;
      if (id === null) {
        if (parsed.error) {
          const err = new FsctError(parsed.error.code, parsed.error.message);
          this.emit('error', err);
          this.failAll(err);
          this.socket.destroy();
        }
        return;
      }

      const cb = this.pending.get(id);
      if (!cb) return;
      this.pending.delete(id);
      if (cb.timer) clearTimeout(cb.timer);

      if (parsed.error) {
        cb.reject(new FsctError(parsed.error.code, parsed.error.message));
      } else {
        cb.resolve(parsed.result ?? null);
      }
    } else if (isRpcNotification(parsed)) {
      if (parsed.method === 'device_changed') {
        this.dispatchDeviceChanged(parsed.params);
      } else {
        console.warn('[fsct-client] Unknown notification method:', parsed.method);
      }
    }
  }

  private dispatchDeviceChanged(params: unknown): void {
    if (typeof params !== 'object' || params === null) {
      console.warn('[fsct-client] Malformed device_changed notification: params not an object');
      return;
    }
    const p = params as Record<string, unknown>;
    const event = p['event'];
    const deviceId = p['device_id'];

    if ((event === 'added' || event === 'removed') && typeof deviceId === 'string') {
      const changeEvent: DeviceChangeEvent = { event, deviceId };
      this.emit('deviceChanged', changeEvent);
    } else {
      console.warn('[fsct-client] Malformed device_changed notification: invalid fields');
    }
  }

  private onError(err: Error): void {
    this.rl.close();
    this.socket.destroy();
    this.failAll(err);
    this.emit('error', err);
  }

  private onClose(): void {
    this.rl.close();
    this.failAll(new Error('IPC connection closed'));
    this.emit('close');
  }

  /** Destroy the underlying socket, rejecting all pending calls. */
  destroy(): void {
    this.rl.close();
    this.failAll(new Error('IPC connection closed'));
    this.socket.destroy();
  }

  private failAll(err: Error): void {
    if (this.#destroyed) return;
    this.#destroyed = true;
    for (const cb of this.pending.values()) {
      if (cb.timer) clearTimeout(cb.timer);
      cb.reject(err);
    }
    this.pending.clear();
  }
}
